ogre2/src/Ogre2SegmentationMaterialSwitcher.cc

```diff
@@ -0,0 +1,321 @@
+
+      VisualPtr visual;
+      try
+      {
+        visual = this->scene->VisualById(visualId);
+      }
+      catch(Ogre::Exception &e)
+      {
+        ignerr << "Ogre Error:" << e.getFullDescription() << "\n";
+      }
+      Ogre2VisualPtr ogreVisual = std::dynamic_pointer_cast<Ogre2Visual>(
+        visual);
+
+      // get class user data
+      Variant labelAny = ogreVisual->UserData("label");
+      int label;
+      try
+      {
+        label = std::get<int>(labelAny);
+      }
+      catch(std::bad_variant_access &e)
+      {
+        // items with no class are considered background
+        label = this->backgroundLabel;
+      }
+
+      // sub item custom parameter to set the pixel color material
+      Ogre::Vector4 customParameter;
+
+      // Material Switching
+      if (this->type == SegmentationType::ST_SEMANTIC)
+      {
+        if (this->isColoredMap)
+        {
+          // semantic material (each pixel has item's color)
+          math::Color color = this->LabelToColor(label);
+          customParameter = Ogre::Vector4(
+            color.R(), color.G(), color.B(), 1.0);
+        }
+        else
+        {
+          // labels ids material (each pixel has item's label)
+          float labelColor = label / 255.0;
+          customParameter = Ogre::Vector4(
+            labelColor, labelColor, labelColor, 1.0);
+        }
+      }
+      else if (this->type == SegmentationType::ST_PANOPTIC)
+      {
+        auto itemName = visual->Name();
+        std::string parentName = this->TopLevelModelVisual(visual)->Name();
+
+        auto it = this->instancesCount.find(label);
+        if (it == this->instancesCount.end())
+          it = this->instancesCount.insert(std::make_pair(label, 0)).first;
+
+        // Multi link model has many links with the same first name and should
+        // have the same pixels color
+        bool isMultiLink = false;
+        if (parentName == prevParentName)
+        {
+          isMultiLink = true;
+        }
+        else
+        {
+          it->second++;
+          prevParentName = parentName;
+        }
+
+        int instanceCount = it->second;
+
+        if (this->isColoredMap)
+        {
+          // convert 24 bit number to int64
+          int compositeId = label * 256 * 256 + instanceCount;
+
+          math::Color color;
+          if (label == this->backgroundLabel)
+            color = this->LabelToColor(label, isMultiLink);
+          else
+            color = this->LabelToColor(compositeId, isMultiLink);
+
+          customParameter = Ogre::Vector4(
+            color.R(), color.G(), color.B(), 1.0);
+        }
+        else
+        {
+          // 256 => 8 bits .. 255 => color percentage
+          float labelColor = label / 255.0;
+          float instanceColor1 = (instanceCount / 256) / 255.0;
+          float instanceColor2 = (instanceCount % 256) / 255.0;
+
+          customParameter = Ogre::Vector4(
+            instanceColor2, instanceColor1, labelColor, 1.0);
```

Comment:
I think when we save the output image in the `ign-sensors::SegmentationCameraSensor` this will produce an Image with channels in the same order `(instanceColor2, instanceColor1, labelColor)` , and we don't want that, since the famous format of the panoptic map is to have the order `(labelColor, instanceColor1, instanceColor2)`

so why is that reversing ?

Reply:
It was reversed as part of the changes made for `ogre2.2`. If I recall correctly, tests were failing before this change.

> the famous format of the panoptic map is to have the order `(labelColor, instanceColor1, instanceColor2)`

Are you saying that the label should be in the first channel, and the instance count should be the last two channels? If so, I think that the implementation you originally wrote is reverse. Looking at the integration test, the label is the last channel, and the instance count is the first 2 channels: https://github.com/AmrElsersy/ign-rendering/blob/segmentation/test/integration/segmentation_camera.cc#L215-L223

It's worth mentioning that although the order of the parameters were reversed for `Ogre2.2`, the code segment I linked above which checks for label/instance count did not change, so I believe that functionality is still the same.

Reply:
Also, looking at your tutorial in `ign-sensors`, I see that the label is the last channel, and the instance count is in the first two channels: https://github.com/ignitionrobotics/ign-sensors/blob/7f8f045a7f39a13fccdb847b0aa55d70423cc3db/tutorials/05_segmentation_camera.md#processing-the-segmentation-sensor-via-ign-transport

So, I think that the change we had to make for Ogre2.2 is okay since this behavior of label in the last channel and instance count in first two channels is preserved

Reply:
Yeah, It was reversed via the C++ API (when we loop over pixels), but when we get the image buffer and save it locally, the saved image was in the correct order (in the order of the parameters we passed `(labelColor, instanceColor1, instanceColor2)`) ... the order of the C++ API dosn't really matter, it is ok to be any order 

But ok I will check the order in the saved images again

Reply:
yeah it was reversed as the test was expecting the label to be in the 3rd channel. Same thing was done to the bounding camera when converting it to ogre 2.2. I never ran the tests before the conversion though. so I don't know if it was passing or not. How about adding a brief comment documenting what each component in the pixel is in the public API: https://github.com/ignitionrobotics/ign-rendering/blob/4ed1b40b0985b20e43388506080f1e268520e5d8/include/ignition/rendering/SegmentationCamera.hh#L62?

Reply:
> the test was expecting the label to be in the 3rd channel. Same thing was done to the bounding camera when converting it to ogre 2.2. I never ran the tests before the conversion though. so I don't know if it was passing or not.

Yes, you're right about the test expecting the label to be in the 3rd channel. The tests were passing before the conversion to ogre 2.2, so the changes you made for the ogre 2.2 conversion are okay :+1: 

> How about adding a brief comment documenting what each component in the pixel is in the public API?

Done: https://github.com/ignitionrobotics/ign-rendering/pull/329/commits/3e3711bd49c6641a5a78247eac71a93676e0481b